pkg/openfeature/event_executor.go

```diff
@@ -0,0 +1,381 @@
+func (e *eventExecutor) startListeningAndShutdownOld(newProvider providerReference, oldReference providerReference) error {
+
+	// check if this provider already actively handled - 1:N binding capability
+	if !isRunning(newProvider, e.activeSubscriptions) {
+		e.activeSubscriptions = append(e.activeSubscriptions, newProvider)
+
+		go func() {
+			v, ok := newProvider.featureProvider.(EventHandler)
+			if !ok {
+				return
+			}
+
+			// event handling of the new feature provider
+			for {
+				select {
+				case event := <-v.EventChannel():
+					e.eventChan <- eventPayload{
+						event:   event,
+						handler: newProvider.featureProvider,
+					}
+				case <-newProvider.shutdownSemaphore:
+					return
+				}
+			}
+		}()
+	}
+
+	// shutdown old provider handling
+
+	// check if this provider is still bound - 1:N binding capability
+	if isBound(oldReference, e.defaultProviderReference, maps.Values(e.namedProviderReference)) {
+		return nil
+	}
+
+	// drop from active references
+	for i, r := range e.activeSubscriptions {
+		if r == oldReference {
+			e.activeSubscriptions = append(e.activeSubscriptions[:i], e.activeSubscriptions[i+1:]...)
+		}
+	}
+
+	_, ok := oldReference.featureProvider.(EventHandler)
+	if !ok {
+		// no shutdown for non event handling provider
+		return nil
+	}
+
+	// avoid shutdown lockouts
+	select {
+	case oldReference.shutdownSemaphore <- "":
+		return nil
+	case <-time.After(200 * time.Millisecond):
+		return fmt.Errorf("old event handler %s timeout waiting for handler shutdown",
+			oldReference.featureProvider.Metadata().Name)
+	}
+}
+
+// startEventListener trigger the event listening of this executor
+func (e *eventExecutor) startEventListener() {
+	e.once.Do(func() {
+		go func() {
+			for payload := range e.eventChan {
+				e.triggerEvent(payload.event, payload.handler)
+			}
+		}()
+	})
+}
+
+// triggerEvent performs the actual event handling
+func (e *eventExecutor) triggerEvent(event Event, handler FeatureProvider) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	// first run API handlers
+	for _, c := range e.apiRegistry[event.EventType] {
+		e.executeHandler(*c, event)
+	}
+
+	// then run client handlers
+	var nameAssociations []string
+	for name, reference := range e.namedProviderReference {
+		if reference.featureProvider == handler {
+			nameAssociations = append(nameAssociations, name)
+		}
+	}
+
+	for _, nameAssociation := range nameAssociations {
+		for _, c := range e.scopedRegistry[nameAssociation].callbacks[event.EventType] {
+			e.executeHandler(*c, event)
+		}
+	}
+
+	if e.defaultProviderReference.featureProvider != handler {
+		return
+	}
+
+	// handling the default provider - invoke default provider bound (no provider associated) handlers by filtering
+	var defaultHandlers []EventCallback
+
+	for clientName, registry := range e.scopedRegistry {
+		if _, ok := e.namedProviderReference[clientName]; !ok {
+			defaultHandlers = append(defaultHandlers, registry.callbacks[event.EventType]...)
```

Comment:
probably we could also do the same here

Reply:
Update with the latest commit - https://github.com/open-feature/go-sdk/pull/190/files#diff-5088e3f82f28056163763bfb774c1b8e91bb585580b4b1add90b133445b8bae6R322-R330